utils/buffer/examples/buffer_pool_exhaustion.rs

```diff
@@ -0,0 +1,40 @@
+// # Handling Buffer Pool Exhaustion and Heap Allocation
+//
+// 1. Write bytes to fill up back of buffer pool.
+// 2. Expect pool mode will switch to front on next write.
+// 3. Write bytes to full up front of buffer pool.
+// 4. Expect pool mode will switch to alloc on next write.
+// 5. Write bytes bigger than buffer pool capacity so that it is allocated on the heap.
+
+use buffer_sv2::{Buffer, BufferPool};
+
+fn main() {
+    let mut buffer_pool = BufferPool::new(8); // 8 bytes capacity
+
+    // Write data to the buffer and retrieve it
+    let write_and_retrieve = |pool: &mut BufferPool<_>, data: &[u8]| {
+        let writable = pool.get_writable(data.len());
+        writable.copy_from_slice(data);
+        println!("Buffer Pool: {:?}", &pool);
+        let data_slice = pool.get_data_by_ref(data.len());
+        println!("Slice from Buffer Pool: {:?}", data_slice.as_ref());
+        // data_slice goes out of scope here
+    };
+
+    let data_bytes = b"abcd"; // 4 bytes
+    println!("DATA: {:?}", data_bytes);
+    write_and_retrieve(&mut buffer_pool, data_bytes);
+    assert!(&buffer_pool.is_back_mode());
+    println!("");
+
+    let data_bytes = b"efg"; // 3 bytes
+    println!("DATA: {:?}", data_bytes);
+    write_and_retrieve(&mut buffer_pool, data_bytes);
+    assert!(&buffer_pool.is_front_mode());
```

Comment:
Fails here as the pool is still in back mode.

```sh
DATA: [97, 98, 99, 100]
Buffer Pool: BufferPool { pool_back: PoolBack { back_start: 0, len: 0 }, mode: Back, shared_state: SharedState(0), inner_memory: InnerMemory { pool: [97, 98, 99, 100, 0, 0, 0, 0], raw_offset: 0, raw_len: 4, slots: [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0)], len: 0 }, system_memory: BufferFromSystemMemory { inner: [], cursor: 0, start: 0 }, start: 0 }
Slice from Buffer Pool: [97, 98, 99, 100]

DATA: [101, 102, 103]
Buffer Pool: BufferPool { pool_back: PoolBack { back_start: 0, len: 0 }, mode: Back, shared_state: SharedState(0), inner_memory: InnerMemory { pool: [97, 98, 99, 100, 101, 102, 103, 0], raw_offset: 0, raw_len: 7, slots: [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0)], len: 0 }, system_memory: BufferFromSystemMemory { inner: [], cursor: 0, start: 0 }, start: 0 }
Slice from Buffer Pool: [97, 98, 99, 100, 101, 102, 103]
thread 'main' panicked at buffer/examples/buffer_pool_exhaustion.rs:33:5:
assertion failed: &buffer_pool.is_front_mode()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

@fi3, I set my buffer pool capacity to 8 bytes. I wrote 4 bytes to the back. Then I wrote 3 bytes, which I would expect to be written to the front of the pool, since we have the back half of the buffer filled with the first write. Instead, the pool is still in back mode during this second write. Then I write four more bytes, such that the 8 byte initialized capacity is exceeded by 4 bytes, and the pool mode switches to alloc mode, as expected.

Why is front mode never activated? 

In the example, [`PoolBuffer.get_writable`](https://github.com/stratum-mining/stratum/blob/b8e77156a8a368bcbfe37fe74b5a2a24b8474b04/utils/buffer/src/buffer_pool/mod.rs#L580) is called, which calls [`PoolBack.get_writable`](https://github.com/stratum-mining/stratum/blob/b8e77156a8a368bcbfe37fe74b5a2a24b8474b04/utils/buffer/src/buffer_pool/pool_back.rs#L146).  At the beginning of the `PoolBack.get_writable`, whether or not the pool has byte capacity is checked with a call to `InnerMemory::has_tail_capacity`(https://github.com/stratum-mining/stratum/blob/b8e77156a8a368bcbfe37fe74b5a2a24b8474b04/utils/buffer/src/buffer_pool/mod.rs#L265). 

If the data trying to be written to the buffer is less than the total buffer capacity the buffer pool was initialized with (8 bytes), then the `get_writable` method returns true, keeping it in back mode.

If the data exceeds the capacity, then we check if the tail is clearable. The tail is not clearable in this example because we have not removed any of the data we have written to the buffer yet. So the `get_writable` returns an [`Err(PoolMode::Alloc)`](https://github.com/stratum-mining/stratum/blob/b8e77156a8a368bcbfe37fe74b5a2a24b8474b04/utils/buffer/src/buffer_pool/pool_back.rs#L162).

Can you help me understand why we don't switch to front mode in this example? Is it the way I have set it up? What case would we proceed past the `tail_is_clearable` check in the `PoolBack::get_writable` and hit the `match` statement? Even in this match statement, the `PoolMode::Front` is ever set.

Reply:
The Buffer Pool in the above example has a size of 8 bytes, with each slot containing 1 byte of memory capacity. In the example, we are not consuming the slices, so they persist and eventually fill the buffer.

This example clearly demonstrates the transition from back mode to front mode and then to alloc mode.
```
use std::collections::VecDeque;

use buffer_sv2::{Buffer, BufferPool};

fn main() {
    let mut buffer_pool = BufferPool::new(8); // 8 bytes capacity

    let mut slices = VecDeque::new();
    let data_bytes = b"a";
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);  

    let data_bytes = b"e";
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);  
 
    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    assert!(&buffer_pool.is_back_mode());

    slices.pop_front();

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);
    assert!(&buffer_pool.is_front_mode());

    let data_bytes = b"h"; // 1 bytes
    let writable = buffer_pool.get_writable(data_bytes.len());
    writable.copy_from_slice(data_bytes);
    let data_slice = buffer_pool.get_data_owned();  
    slices.push_back(data_slice);

    assert!(&buffer_pool.is_alloc_mode());
    
}

```

Reply:
THANK YOU!